twilio-keycloak-provider/lib/src/main/java/org/philanthropydatacommons/auth/twilio/authenticator/SmsAuthenticator.java

```diff
@@ -0,0 +1,125 @@
+import org.philanthropydatacommons.auth.twilio.Sms;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.ws.rs.core.Response;
+import java.util.Locale;
+public class SmsAuthenticator implements Authenticator {
+    private static final Logger LOGGER = LoggerFactory.getLogger(SmsAuthenticator.class);
+    private static final String TPL_CODE = "login-sms.ftl";
+
+    static {
+        LOGGER.debug("Initializing the SmsAuthenticator class.");
+    }
+
+    @Override
+    public void authenticate(AuthenticationFlowContext context) {
+        AuthenticatorConfigModel config = context.getAuthenticatorConfig();
+        KeycloakSession session = context.getSession();
+        UserModel user = context.getUser();
+
+        String mobileNumber = user.getFirstAttribute("mobile_number");
+        // mobileNumber of course has to be further validated on proper format, country code, ...
+
+        int length = Integer.parseInt(config.getConfig().get("length"));
+        int ttl = Integer.parseInt(config.getConfig().get("ttl"));
+
+        String code = SecretGenerator.getInstance().randomString(length, SecretGenerator.DIGITS);
+        AuthenticationSessionModel authSession = context.getAuthenticationSession();
+        authSession.setAuthNote("code", code);
+        authSession.setAuthNote("ttl", Long.toString(System.currentTimeMillis() + (ttl * 1000L)));
+
+        try {
+            Theme theme = session.theme().getTheme(Theme.Type.LOGIN);
+            Locale locale = session.getContext().resolveLocale(user);
+            String smsAuthText = theme.getMessages(locale).getProperty("smsAuthText");
+            String smsText = String.format(smsAuthText, code, Math.floorDiv(ttl, 60));
+
+            Sms sms = new Sms();
+            sms.send(mobileNumber, smsText);
+            LOGGER.debug("Sent OTP via SMS to user '{}'", user.getId());
+
+            context.challenge(context.form().setAttribute("realm", context.getRealm()).createForm(TPL_CODE));
+        } catch (Exception e) {
+            context.failureChallenge(AuthenticationFlowError.INTERNAL_ERROR,
+                    context.form().setError("smsAuthSmsNotSent", e.getMessage())
+                            .createErrorPage(Response.Status.INTERNAL_SERVER_ERROR));
+        }
+    }
+
+    @Override
+    public void action(AuthenticationFlowContext context) {
+        String enteredCode = context.getHttpRequest().getDecodedFormParameters().getFirst("code");
+
+        AuthenticationSessionModel authSession = context.getAuthenticationSession();
+        String code = authSession.getAuthNote("code");
+        String ttl = authSession.getAuthNote("ttl");
+
+        if (code == null || ttl == null) {
+            context.failureChallenge(AuthenticationFlowError.INTERNAL_ERROR,
+                    context.form().createErrorPage(Response.Status.INTERNAL_SERVER_ERROR));
+            return;
+        }
+
+        boolean isValid = enteredCode.equals(code);
```

Comment:
I believe this is vulnerable to a [timing attack](https://codahale.com/a-lesson-in-timing-attacks/): `String#equals` will do a character-by-character comparison, and return on the first inequality; the time it takes to do so can reveal information to an attacker about the correct value.

It may be the case (and I hope it is) that Keycloak has built in defenses against brute force attacks, but in the interest of defense-in-depth, we should use a constant time comparison: [`MessageDigest#isEqual`](https://docs.oracle.com/javase/8/docs/api/java/security/MessageDigest.html#isEqual-byte:A-byte:A-)

```suggestion
        boolean isValid = MessageDigest.isEqual(
            code.getBytes(StandardCharsets.UTF_8),
            enteredCode.getBytes(StandardCharsets.UTF_8)
        );
```

I'd also suggest doing the expiration time check before the code check, both because it is cheaper and to reveal as little to an attacker as possible.

Reply:
I need to still apply this change, but it is probably worthwhile to report this issue upstream because here it is: https://github.com/dasniko/keycloak-2fa-sms-authenticator/blob/6e5765bbf740afad5df0d48ddf8302cfb0e0f4a0/src/main/java/dasniko/keycloak/authenticator/SmsAuthenticator.java#L73

Reply:
I applied both changes. Thanks for finding this!